From: Kefeng Wang <wangkefeng.wang@huawei.com>
Subject: mm: hwpoison: support recovery from ksm_might_need_to_copy()
Date: Fri, 9 Dec 2022 15:28:01 +0800

When the kernel copies a page from ksm_might_need_to_copy(), but runs into
an uncorrectable error, it will crash since poisoned page is consumed by
kernel, this is similar to Copy-on-write poison recovery.

When an error is detected during the page copy, return VM_FAULT_HWPOISON
in do_swap_page(), and install a hwpoison entry in unuse_pte() when
swapoff, which help us to avoid system crash.  Note, memory failure on a
KSM page will be skipped, but still call memory_failure_queue() to be
consistent with general memory failure process.

We are trying to support ARCH_HAS_COPY_MC on arm64[1] and trying to
recover from CoW faults[2], also Tony wants the same thing (recover
from CoW) on X86[3].  The kernel copy in ksm_might_need_to_copy() could
recover, this is an enhance of COPY_MC, so I think no need to add Fixes
and stable.

[1] 
https://lore.kernel.org/linux-arm-kernel/20220812070557.1028499-1-tongtiangen@huawei.com/
[2] 
https://lore.kernel.org/linux-arm-kernel/20220812070557.1028499-5-tongtiangen@huawei.com/
[3] 
https://lore.kernel.org/lkml/20221031201029.102123-2-tony.luck@intel.com/

Link: https://lkml.kernel.org/r/20221209072801.193221-1-wangkefeng.wang@huawei.com
Signed-off-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Miaohe Lin <linmiaohe@huawei.com>
Cc: Naoya Horiguchi <naoya.horiguchi@nec.com>
Cc: Tony Luck <tony.luck@intel.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/ksm.c      |    8 ++++++--
 mm/memory.c   |    3 +++
 mm/swapfile.c |    2 +-
 3 files changed, 10 insertions(+), 3 deletions(-)

--- a/mm/ksm.c~mm-hwposion-support-recovery-from-ksm_might_need_to_copy
+++ a/mm/ksm.c
@@ -2629,8 +2629,12 @@ struct page *ksm_might_need_to_copy(stru
 		new_page = NULL;
 	}
 	if (new_page) {
-		copy_user_highpage(new_page, page, address, vma);
-
+		if (copy_mc_user_highpage(new_page, page, address, vma)) {
+			put_page(new_page);
+			new_page = ERR_PTR(-EHWPOISON);
+			memory_failure_queue(page_to_pfn(page), 0);
+			return new_page;
+		}
 		SetPageDirty(new_page);
 		__SetPageUptodate(new_page);
 		__SetPageLocked(new_page);
--- a/mm/memory.c~mm-hwposion-support-recovery-from-ksm_might_need_to_copy
+++ a/mm/memory.c
@@ -3840,6 +3840,9 @@ vm_fault_t do_swap_page(struct vm_fault
 		if (unlikely(!page)) {
 			ret = VM_FAULT_OOM;
 			goto out_page;
+		} else if (unlikely(PTR_ERR(page) == -EHWPOISON)) {
+			ret = VM_FAULT_HWPOISON;
+			goto out_page;
 		}
 		folio = page_folio(page);
 
--- a/mm/swapfile.c~mm-hwposion-support-recovery-from-ksm_might_need_to_copy
+++ a/mm/swapfile.c
@@ -1768,7 +1768,7 @@ static int unuse_pte(struct vm_area_stru
 
 	swapcache = page;
 	page = ksm_might_need_to_copy(page, vma, addr);
-	if (unlikely(!page))
+	if (IS_ERR_OR_NULL(page))
 		return -ENOMEM;
 
 	pte = pte_offset_map_lock(vma->vm_mm, pmd, addr, &ptl);
_
